### PR TITLE
[ACI-2290] Xcode DerivedData cache commands

### PR DESCRIPTION
**Pull request by imrekel**

### CODE REVIEW

File: cmd/saveXcodeDerivedData.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+		logger.TInfof("✅ DerivedData saved into Bitrise Build Cache ")
+
+		return nil
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(saveXcodeDerivedDataCmd)
+
+	saveXcodeDerivedDataCmd.Flags().String("key", "", "The cache key to use for the saved cache item (set to the current git branch by default)")
```
Comment (imrekel) [inline cmd/saveXcodeDerivedData.go:47]:
```suggestion
	saveXcodeDerivedDataCmd.Flags().String("key", "", "The cache key to use for the saved cache item (set to the current Bitrise app's slug and git branch by default)")
```

File: cmd/restoreXcodeDerivedData.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+			return fmt.Errorf("get cache key: %w", err)
+		}
+	}
+	logger.Infof("(i) Cache key: %s", cacheKey)
+
+	endpointURL := common.SelectEndpointURL(envProvider("BITRISE_BUILD_CACHE_ENDPOINT"), envProvider)
+	logger.Infof("(i) Build Cache Endpoint URL: %s", endpointURL)
+
+	logger.TInfof("Downloading cache archive for key %s", cacheKey)
+	if err := xcode.DownloadFromBuildCache(cacheArchivePath, cacheKey, authConfig.AuthToken, endpointURL, logger); err != nil {
```
Comment (zsolt-marta-bitrise) [inline cmd/restoreXcodeDerivedData.go:71]: Created this to refactor to use interfaces https://bitrise.atlassian.net/browse/ACI-2397

File: internal/xcode/metadata.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+func RestoreMTime(metadata *Metadata, rootDir string, logger log.Logger) error {
+	updated := 0
+
+	logger.Infof("(i) %d files' info loaded from cache metadata", len(metadata.FileInfos))
+
+	for _, fi := range metadata.FileInfos {
+		path := filepath.Join(rootDir, fi.Path)
+
+		// Skip if file doesn't exist
+		if _, err := os.Stat(path); os.IsNotExist(err) {
```
Comment (zsolt-marta-bitrise) [inline internal/xcode/metadata.go:75]: Let's log (preferably warn) in this case

File: cmd/saveXcodeDerivedData.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	saveXcodeDerivedDataCmd.Flags().String("cache-archive", "bitrise-dd-cache/dd.tar.zst", "Path to the uploadable cache archive with the contents of the DerivedData folder")
+	saveXcodeDerivedDataCmd.Flags().String("project-root", "", "Path to the iOS project folder to be built (this is used when saving the modification time of the source files)")
```
Comment (zsolt-marta-bitrise) [inline cmd/saveXcodeDerivedData.go:49]: Can we default to `"."` in this case?

File: cmd/restoreXcodeDerivedData.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+		return nil
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(restoreXcodeDerivedDataCmd)
+
+	restoreXcodeDerivedDataCmd.Flags().String("key", "", "The cache key to use for the saved cache item (set to the current git branch by default)")
+	restoreXcodeDerivedDataCmd.Flags().String("cache-archive", "bitrise-dd-cache/dd.tar.zst", "Path to the uploadable cache archive with the contents of the DerivedData folder")
+	restoreXcodeDerivedDataCmd.Flags().String("project-root", "", "Path to the iOS project folder to be built (this is used when restoring the modification time of the source files)")
```
Comment (zsolt-marta-bitrise) [inline cmd/restoreXcodeDerivedData.go:47]: Same as [here](https://github.com/bitrise-io/bitrise-build-cache-cli/pull/28/files#r1652454344), can we default to `"."`?

File: internal/xcode/archive.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+func ExtractCacheArchive(fileName string, logger log.Logger) error {
+	envRepo := env.NewRepository()
+	archiver := compression.NewArchiver(
+		logger,
+		envRepo,
+		compression.NewDependencyChecker(logger, envRepo))
+
+	err := archiver.Decompress(fileName, "")
+	if err != nil {
+		return fmt.Errorf("failed to decompress: %w", err)
```
Comment (zsolt-marta-bitrise) [inline internal/xcode/archive.go:44]: nit: For wrapping errors don't add "fail" or "error", that just bloats the stack and makes it hard to read

Comment (zsolt-marta-bitrise) [reply internal/xcode/archive.go:44]: Just do a replace all on "failed to" -> "" everywhere

File: internal/xcode/file_info.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+func calculateFileInfos(rootDir string, logger log.Logger) ([]FileInfo, error) {
+	if rootDir == "" {
+		return nil, fmt.Errorf("missing rootDir")
+	}
+
+	var fileInfos []FileInfo
+
+	// Walk through the directory tree
+	err := filepath.WalkDir(rootDir, func(path string, d fs.DirEntry, err error) error {
```
Comment (zsolt-marta-bitrise) [inline internal/xcode/file_info.go:27]: Please extract to separate method to simplify wrapper/caller

File: internal/xcode/metadata_test.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+	testInputFile, err := os.CreateTemp(testRootDir, "test-file.swift")
+	if err != nil {
+		t.Logf("Error creating temp file: %v", err)
+
+		return
+	}
+	testInputFile.Close()
+
+	tests := []struct {
```
Comment (zsolt-marta-bitrise) [inline internal/xcode/metadata_test.go:33]: Later we should add cases for skipping symbolic links / directories

File: internal/xcode/tracker.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+		"duration_ms": duration.Milliseconds(),
+		"file_count":  fileCount,
+	}
+	t.tracker.Enqueue("step_save_xcode_build_cache_metadata_saved", properties)
+}
+
+func (t *StepAnalyticsTracker) LogArchiveCreated(duration time.Duration, archiveSizeBytes int64) {
+	properties := analytics.Properties{
+		"duration_ms":       duration.Milliseconds(),
+		"archive_size_byte": archiveSizeBytes,
```
Comment (zsolt-marta-bitrise) [inline internal/xcode/tracker.go:41]: should by `archive_size_bytes`

**Comment by imrekel** [comment]
let's continue on https://github.com/bitrise-io/bitrise-build-cache-cli/pull/28

**Pull request closed without merging**